utils/src/lib.rs

```diff
@@ -1,32 +1,261 @@
 use std::{
-    borrow::Cow,
+    fmt::Display,
     time::{Duration, Instant},
 };
 
+use bytesize::ByteSize;
 use memory_stats::{MemoryStats, memory_stats};
-use tracing::info;
+use thousands::Separable;
 
-#[macro_export]
+#[derive(Default)]
+struct AllocatorMetrics {
+    /// Total number of bytes allocated so far.
+    allocated: usize,
 
-macro_rules! info_metrics {
-    ($open: expr, $close: expr $(,)?) => {
-        let _guard = $crate::MeasureStage::new($open.into(), $close.into()).guard();
+    /// Total number of bytes deallocated so far.
+    deallocated: usize,
+
+    /// Number of alloc calls.
+    alloc_calls: usize,
+
+    /// Peak memory usage in bytes.
+    ///
+    /// Note: The peak memory usage can be reset, this is used to measure
+    /// the memore usage in a span of time.
+    peak: usize,
+}
+
+/// Generates memory flame graphs for the period of time this object is alive.
+///
+/// Data collection starts when the struct is created, and it finishes when the
+/// struct is dropped.
+///
+/// Note: Flame graph generation takes a long time, the execution of the drop will
+/// take a long time.
+/// Note: To generate flamegraphs an additional environment variable called `FLAMEGRAPH`
+/// must be present with a non-empty string, the variable's contents will be used
+/// as the generate file prefix.
+///
+/// # Panics
+///
+/// Only one of this structures may exist at any time.
+pub struct MemoryFlameGraph {}
+
+impl MemoryFlameGraph {
+    /// Starts memory flame graph collection
+    ///
+    /// # Panics
+    ///
+    /// If there is already a flame graph being collected.
+    pub fn new() -> Self {
+        track::flame_graph_enable();
+        Self {}
+    }
+}
+
+impl Drop for MemoryFlameGraph {
+    fn drop(&mut self) {
+        track::flame_graph();
+    }
+}
```

Comment:
Now only the flame graph is using the guard style api, the rest uses a init + consume to gather the data for a span, the span is then used to print the results.